devdoc/jdp/jdp-2019-06-ability-to-control-statements-execution.md

```diff
@@ -0,0 +1,101 @@
+# jdp-2019-06: Ability to control statements execution
+
+## Status
+
+- Draft
+
+## Type
+
+- Feature-Specification
+
+## Context
+
+At any given time, a different number of statements `org.firebirdsql.gds.ng.FbStatement` 
+can be executed in an application. The two main states in which a statement 
+can be exist are the execution `org.firebirdsql.gds.ngStatementState.EXECUTING` 
+and fetching `org.firebirdsql.gds.ngStatementState.CURSOR_OPEN`.
+
+The first problem is that these statuses allow us to track the status of the 
+statement, but do not allow it centrally for all statements. Partly, this problem 
+can be solved at the application level. But for this, each statement will have 
+to be wrapped to control executing. At the driver level, tracking is much easier.
+The second problem is the control of fetching rows. Executing or fetching data 
+can take considerable time and at the moment we do not know whether operations 
+are performed or not.
+
+The main idea of ​​this proposal is to add a mechanism that allows at an application
+level to know what operations are performed at a given moment in time.
+
+## Decision
+
+This proposal should be taken as a suggestion and may be revised.
+
+Implementation involves the addition of several interfaces and classes to control 
+the execution of statements.
+
+### Operation interface
+
+At the `org.firebirdsql.gds` package it is proposed to add the `Operation` interface, 
+which contains methods for managing operations. At time of writing, we are only 
+interested in canceling the operation, so the interface contains only the method for 
+canceling a operation `void cancelOperation()`. Other methods may be added later.
+
+### Interface implementation
+
+The interface implementation can be placed at `org.firebirdsql.gds.impl` package.
+A separate class can be implemented for each type of operation. Since we only 
+need to cancel the operation, then object of `GDSHelper` class is suitable for this.
+This class already contains method for cancelling current operation `cancelOperation()`.
+
+### Control the statement executions
+
+The main idea is to inform the client application about the starting and finishing 
+of the statement operation, such as execution and data fetching. At the same time, 
+we should **not overload** the driver, but only **provide way for control**. At the 
```

Comment:
What kind of 'overload' are you thinking of, and how does the design try to prevent it?

Reply:
Here, 'overload' means that driver provides only a some interface for controlling execution and etc. There is no need for such an implementation in the driver itself (or there is a simple implementation).